tools/performance-server/src/main/kotlin/routes/route.kt

```diff
@@ -139,191 +47,281 @@ data class BuildRegister(val buildId: String, val teamCityUser: String, val team
+                        (measurement.field("build.number") eq FieldType.InfluxString(it))).then { dbResponse ->
+                    val measurements = (dbResponse as List<BenchmarkMeasurement>)
+                    val report = measurements.toReport()
+                    val failuresNumber = report?.let { report ->
+                        val summaryReport = SummaryBenchmarksReport(report)
+                        summaryReport.failedBenchmarks.size
+                    } ?: 0
+                    val buildDescription = measurements.firstOrNull()?.let {
+                        Build(it.buildNumber!!.value, it.buildStartTime!!.value, it.buildEndTime!!.value, it.buildBranch!!.value,
+                                it.buildCommits!!.value, request.params.type, failuresNumber)
+                    }
+                    buildDescription
+                }.catch {
+                    response.sendStatus(400)
+                }
+            }
+
+            Promise.all(responseLists.toTypedArray()).then { resultList ->
+                response.json(resultList)
+            }.catch {
+                response.sendStatus(400)
+            }
+        }.catch {
+            response.sendStatus(400)
+        }
     })
 
-    router.get("/builds/:target/:type/:branch", { request, response ->
-        val builds = LocalCache[request.params.target]
-        response.json(prepareBuildsResponse(builds, request.params.type, request.params.branch))
+    // Get values of current metric.
+    router.get("/metricValue/:target/:type/:metric", { request, response ->
+        val measurement = BenchmarkMeasurement()
+        val metric = request.params.metric
+        val target = request.params.target.toString().replace('_', ' ').asDynamic()
+        var samples: List<String>? = null
+        var agregation = "geomean"
+        var normalize = false
+        var branch: String? = null
+
+        // Parse parameters from request if it exists.
+        if (request.query != undefined) {
+            if (request.query.samples != undefined) {
+                samples = request.query.samples.toString().split(",").map { it.trim() }
+            }
+            if (request.query.agr != undefined) {
+                agregation = request.query.agr.toString()
+            }
+            if (request.query.normalize != undefined) {
+                normalize = true
+            }
+            if (request.query.branch != undefined) {
+                branch = request.query.branch
+            }
+        }
+
+        // Get golden results to normalize data.
+        val golden = GoldenResultMeasurement()
+        val goldenResults = golden.select(golden.all()).then { results ->
+            val parsedNormalizeResults = mutableMapOf<String, MutableMap<String, Double>>()
+            (results as List<GoldenResultMeasurement>).forEach {
+                parsedNormalizeResults.getOrPut(it.benchmarkName!!,
```

Comment:
Shall benchmarkName and benchmarkMetric be nullable then?

Reply:
Yes, they should, because Tag should provide delegate a nullable value.